src/test/ui/closures/2229_closure_analysis/move_closure.rs

```diff
@@ -156,16 +154,54 @@ fn truncate_box_derefs() {
     //~^ ERROR: First Pass analysis includes:
     //~| ERROR: Min Capture analysis includes:
         println!("{}", t.1.0);
-        //~^ NOTE: Capturing t[(1, 0),Deref,(0, 0)] -> ByValue
+        //~^ NOTE: Capturing t[(1, 0),Deref,(0, 0)] -> ImmBorrow
         //~| NOTE: Min Capture t[(1, 0)] -> ByValue
     };
 }
 
+struct Foo { x: i32 }
+
+// Ensure that even in move closures, if the data is not owned by the root variable
+// then we don't truncate the derefs or a ByValue capture, rather do a reborrow
+fn box_mut_1() {
+    let mut foo = Foo { x: 0 } ;
+
+    let p_foo = &mut foo;
+    let box_p_foo = Box::new(p_foo);
+
+    let c = #[rustc_capture_analysis] move || box_p_foo.x += 10;
+    //~^ ERROR: attributes on expressions are experimental
+    //~| NOTE: see issue #15701 <https://github.com/rust-lang/rust/issues/15701>
+    //~| First Pass analysis includes:
+    //~| NOTE: Capturing box_p_foo[Deref,Deref,(0, 0)] -> UniqueImmBorrow
+    //~| Min Capture analysis includes:
+    //~| NOTE: Min Capture box_p_foo[Deref,Deref,(0, 0)] -> UniqueImmBorrow
```

Comment:
This is a "pre-existing bug", but uniqueImmBorrow actually seems wrong here. I think this should be a Mut borrow. Where does that come about?

Reply:
Yea I thought this should be a mut borrow as well

Reply:
Right, so you highlighted this code in another thread, but I think you meant it for here:

https://github.com/rust-lang/rust/blob/efeea79a6bb6867b220471e0396b82b6b68ebc4c/compiler/rustc_typeck/src/check/upvar.rs#L1495-L1503

The problem here is that this code is still oriented around a single variable, and not a place.

Reply:
This is actually a bit tricky =)

Reply:
Given `foo: &mut (u32, u32)`, there are a few scenarios to consider:

* If the closure mutates `foo.0` but we wind up capturing `foo`, we only need a "unique" borrow of `foo`.
* If the closure mutates `foo.0` *and we capture `foo.0`*, we need a "mut" borrow.

In short, the need for a "mut" borrow depends on what data we ultimately capture. 

Let's put off fixing this for a follow-up PR.

Reply:
Filed https://github.com/rust-lang/project-rfc-2229/issues/56